JavaExamples.java

```diff
@@ -1,84 +1,86 @@
 /// Java Examples
 import java.util.*;
+import qual.Immutable;
 import qual.Mutable;
 
-// Mutating immut set, `s` is immut
-void foo(Set<String> s) {
-  Set<String> new_s = s;
-  @Mutable Set<String> new_s1 = s; // ERROR
-  new_s.add("x"); // ERROR
-}
+public class JavaExamples {
+    // Mutating immut set, `s` is immut
+    void foo(@Immutable Set<String> s) {
+        Set<String> new_s = s; // Flow sensitive will refine this to @Immutable
+        @Mutable Set<String> new_s1 = s; // ERROR, type incompatible
+        new_s.add("x"); // ERROR
+    }
 
-// Mutating immut set, `new_s` is immut
-void foo1(@Mutable Set<String> s) {
-  Set<String> new_s = new HashSet<>(s);
-  Set<String> new_s1 = s;
-  new_s.add("x"); // ERROR
-  new_s1.add("x"); //OK
-}
+    // Mutating immut set, `new_s` is immut
+    void foo1(@Mutable Set<String> s) {
+        @Immutable Set<String> new_s = new @Immutable HashSet<>(s);
+        Set<String> new_s1 = s; // Flow sensitive will refine this to @Mutable
+        new_s.add("x"); // ERROR
+        new_s1.add("x"); //OK
+    }
 
-// Mutating mut set
-void foo2(Set<String> s) {
-  @Mutable Set<String> new_s = new HashSet<>(s);
-  new_s.add("x"); // OK
-}
+    // Mutating mut set
+    void foo2(Set<String> s) {
+        @Mutable Set<String> new_s = new HashSet<>(s);
+        new_s.add("x"); // OK
+    }
+
+    // Type parameter mutability
+    void foo3(Set<List<String>> s, List<String> l) {
+        assert s.get(l) != null;
+        List<String> v = s.get(l);
+        @Mutable List<String> v2 = s.get(l); // ERROR
+        v.add("x"); // ERROR
+    }
+
+    void foo4(Person p) {
+        p.name = "Jenny"; // ERROR, this should be forbid by compiler by adding final before String
+        p.family.add("Jenny"); // ERROR, can not mutate immut list
+    }
 
-// Type parameter mutability
```

Comment:
@txiang61 Could you explain what `foo3` try to do? The method invocation on Set is not right and the comment `type parameter mutability` should related generics, right?

Reply:
In this example I was just trying to say that the items in the `Set` are immutable, because there are no `@Mutable` before `List`. 

Yes need checks on generic type parameters as well. Can you add this to the examples? Thanks!
```
class ImmutSet<T> {}

class @Mutable MutList<@Mutable T> {}

void foo(ImmutSet<MutList<T>> s) {  // ERROR 
}

void foo(MutList<MutList<T>> s) {  // OK 
}
```